Add support for AssumeRoleWithWebIdentity

mwieczorek opened the pull request:
<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

Closes #149 <!--- Pull Requests should have an associated issue or may be closed --->

This PR adds an option to pass webIdentityToken to `Config` and use it for getting credentials. No other 'source credentials' are required.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/aws-sdk-go-base?pullRequest=178)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/aws-sdk-go-base?pullRequest=178)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/aws-sdk-go-base?pullRequest=178).</sub>

mwieczorek commented on credentials.go:
```diff
@@ -153,6 +158,25 @@ Error: %w`, err)
 }
 
+func webIdentityCredentialsProvider(ctx context.Context, awsConfig aws.Config, c *Config) (aws.CredentialsProvider, error) {
+	ar := c.AssumeRole
```
I reused `config.AssumeRole` 'block' here, although not every option is valid for 'assumeRoleWithWebIdentity'.
@gdavison - let me know if it's ok, or if you prefer to have something like `config.WebIdentityAssumeRole`

In https://github.com/hashicorp/terraform-provider-aws/issues/22907#issuecomment-1048371318 you accepted combining both at the provider level, but I don't know how it should be done here.

gdavison replied:
I'm reconsidering that statement 🙂 Since the two operations support different options, it would probably be better to explicitly choose one or the other, both here and in the provider.

mwieczorek replied:
ok, I changed the Config and created a separate struct for AssumeRoleWithWebIdentity

mwieczorek commented on internal/config/config.go:
```diff
@@ -79,3 +80,7 @@ func (c Config) ResolveSharedCredentialsFiles() ([]string, error) {
 	return v, nil
 }
+
+func (c Config) GetIdentityToken() ([]byte, error) {
```
I need it for https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/credentials@v1.10.0/stscreds#NewWebIdentityRoleProvider / https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/credentials@v1.10.0/stscreds#IdentityTokenRetriever

Let me know if you prefer to have 'implementation' of that interface separately from `Config` type

gdavison replied:
Since it only references fields from `c.AssumeRoleWithWebIdentity`, it would probably be good to move it there

mwieczorek commented on credentials.go:
```diff
@@ -133,6 +133,11 @@ func getCredentialsProvider(ctx context.Context, c *Config) (aws.CredentialsProv
 		return nil, "", fmt.Errorf("loading configuration: %w", err)
 	}
 
+	if c.WebIdentityToken != "" && c.AssumeRole != nil {
```
Let me know if you'd like to additionally validate/error when AssumeRole is nil when WebIdentityToken is passed.

mwieczorek commented:
I also tested it against a real AWS account/provider/role setup with simple code like
```
        ctx := context.Background()
	c := awsbase.Config{
		Region:           "us-east-1",
		WebIdentityToken: "<my id token>",
		AssumeRole: &config.AssumeRole{
			RoleARN: "arn:aws:iam::123456789012:role/role-name",
		},
	}
	cfg, err := awsbase.GetAwsConfig(ctx, &c)
```

djcrabhat commented:
Oh I hit up against this exact problem today. Was gonna make in gitlab CI project to deploy something across a dozen accounts in my org via aliased providers, found I can't share one identity token file via CLI _and_ get each provider to assume a unique role. Guess I'm in the cutting edge of terraform a CD!

gdavison commented on internal/config/config.go:
```diff
@@ -36,6 +36,7 @@ type Config struct {
 	UseDualStackEndpoint           bool
 	UseFIPSEndpoint                bool
 	UserAgent                      UserAgentProducts
+	WebIdentityToken               string
```
The `WebIdentityToken` should be part of the block used to assume the role, so that it's all defined in one place. This will be more important when, e.g. https://github.com/hashicorp/terraform-provider-aws/issues/22728 is implemented.

gdavison replied:
We may also want to allow specifying the token file, in addition to putting it inline

mwieczorek replied:
`WebIdentityToken` is not part of `AssumeRoleWithWebIdentity`. I also added option to pass `WebIdentityTokenFile`.

gdavison commented on credentials_test.go:
```diff
@@ -189,6 +189,39 @@ func TestAWSGetCredentials_sharedCredentialsFile(t *testing.T) {
 	validateCredentialsProvider(credsParam, "accesskey2", "secretkey2", "", sharedConfigCredentialsSource(fileParamName), t)
 }
 
+func TestAWSGetCredentials_webIdentityToken(t *testing.T) {
```
I've created #227 that adds tests for configuring this using environment variables or the shared config files. Those tests can be updated with support for the configuration block in addition to this test

gdavison replied:
The PR #227 has been merged, so if you rebase this branch or merge `main` into this branch, you can pick up those tests

gdavison commented on credentials.go:
```diff
@@ -153,6 +158,25 @@ Error: %w`, err)
+	client := stsClient(awsConfig, c)
+
+	appCreds := stscreds.NewWebIdentityRoleProvider(client, ar.RoleARN, c, func(opts *stscreds.WebIdentityRoleOptions) {
+		opts.RoleSessionName = ar.SessionName
```
FYI `stscreds.WebIdentityRoleProvider` doesn't currently support the `Policy` field from `sts.AssumeRoleWithWebIdentityInput`. I've created https://github.com/aws/aws-sdk-go-v2/issues/1662 and may create a PR in the SDK

gdavison reviewed (commented):
Thanks for creating this, @mwieczorek. It's looking good. I have a few comments so far.

mwieczorek commented:
Hi @gdavison , thank you for the review. I updated the PR, please take a look.

gdavison commented on credentials_test.go:
```diff
@@ -189,6 +189,39 @@ func TestAWSGetCredentials_sharedCredentialsFile(t *testing.T) {
+		t.Fatalf("unexpected '%[1]T' error getting credentials provider: %[1]s", err)
+	}
+
+	if a, e := source, ""; a != e {
+		t.Errorf("Expected initial source to be %q, %q given", e, a)
+	}
```
For the other credentials providers, we return the name of the credentials source. In this case, it would be `stscreds.WebIdentityProviderName` (i.e. `WebIdentityCredentials`)

gdavison reviewed (commented):
It's looking good

mwieczorek commented:
hi @gdavison 
I moved `GetIdentityToken` as you suggested, and changed `credentialsProvider` name returned from `getCredentialsProvider`

I also rebased this PR, and updated test cases in TestAssumeRoleWithWebIdentity. I changed the input struct (removed `SetConfig`, added `SetConfigTokenFile` - this allows me to have separate test cases when we pass to config token directly or when we pass path to file with token). Let me know what you think.

gdavison reviewed (approved):
This is great, @mwieczorek. Since AWS has added `Duration` and `Policy` to `stscreds.WebIdentityRoleOptions` (aws/aws-sdk-go-v2#1670), I've added them to `AssumeRoleWithWebIdentity`. I've also added the ability to assume another role after using the web identity

mwieczorek commented:
thanks @gdavison for all your help :)
